File: src/routes/numbering.ts
// DEPRECATED v3.20.0: Superseded by FastAPI /api/v1/* — not mounted in server.ts. Safe to delete after v3.21.0.

/**
 * Project      : SMRITI Retail OS
 * Author       : Jawahar Ramkripal Mallah
 * Designation  : Chief Systems Architect & Creator
 * Email        : support@smritibooks.com
 * Websites     : smritibooks.com | erpnbook.com | aitdl.com
 * Version      : 3.16.0
 * Created      : 2026-07-12
 * Modified     : 2026-07-13
 * Copyright    : © SMRITIBooks.com. All Rights Reserved.
 * License      : Proprietary Commercial Software
 */

import express from "express";
import { pool } from "../db/pool.js";

const router = express.Router();

router.get("/api/numbering/series", async (req, res) => {
  try {
    const dbRes = await pool.query(
      `SELECT id, name, document_type AS "documentType", module, prefix, suffix, 
              running_length AS "runningLength", reset_rule AS "resetRule", 
              current_number AS "currentNumber", last_reset_key AS "lastResetKey", 
              financial_year AS "financialYear", company_code AS "companyCode", 
              mode, description, is_active AS "isActive" 
       FROM document_series 
       WHERE is_deleted = false`
    );
    res.json(dbRes.rows);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

router.post("/api/numbering/series", async (req, res) => {
  const data = req.body;
  const id = "SER-" + Date.now();
  const currentNumber = parseInt(data.currentNumber) || 0;
  const runningLength = parseInt(data.runningLength) || 6;
  const isActive = data.isActive !== false;
  const companyCode = data.companyCode || "SMRITI_IND";
  const financialYear = data.financialYear || "2026-2027";
  const mode = data.mode || "Auto";
  const resetRule = data.resetRule || "Financial Year";

  try {
    await pool.query(
      `INSERT INTO document_series (
        id, uuid, name, document_type, module, prefix, suffix, 
        running_length, reset_rule, current_number, last_reset_key, 
        financial_year, company_code, mode, description, is_active, is_deleted, created_at, modified_at
      ) VALUES ($1, uuid_generate_v4()::varchar, $2, $3, $4, $5, $6, $7, $8, $9, null, $10, $11, $12, $13, $14, false, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
      [
        id,
        data.name,
        data.documentType,
        data.module || null,
        data.prefix || "",
        data.suffix || "",
        runningLength,
        resetRule,
        currentNumber,
        financialYear,
        companyCode,
        mode,
        data.description || null,
        isActive
      ]
    );

    const logId = "NAL-" + Date.now().toString() + Math.floor(Math.random() * 1000);
    const operator = (req.headers["x-user-name"] as string) || "Admin";
    await pool.query(
      `INSERT INTO numbering_audit_logs (
        id, uuid, series_id, series_name, action, document_no, old_value, new_value, details, operator, is_active, is_deleted, created_at, modified_at
      ) VALUES ($1, uuid_generate_v4()::varchar, $2, $3, 'CREATE', '-', '-', $4, $5, $6, true, false, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
      [
        logId,
        id,
        data.name,
        String(currentNumber),
        `Created new Document Series '${data.name}' for ${data.documentType}`,
        operator
      ]
    );

    res.json({
      id,
      name: data.name,
      documentType: data.documentType,
      module: data.module || null,
      prefix: data.prefix || "",
      suffix: data.suffix || "",
      runningLength,
      resetRule,
      currentNumber,
      lastResetKey: null,
      financialYear,
      companyCode,
      mode,
      description: data.description || null,
      isActive
    });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

router.put("/api/numbering/series/:id", async (req, res) => {
  const { id } = req.params;
  const data = req.body;

  try {
    const checkRes = await pool.query(
      `SELECT id, name, document_type AS "documentType", module, prefix, suffix, 
              running_length AS "runningLength", reset_rule AS "resetRule", 
              current_number AS "currentNumber", last_reset_key AS "lastResetKey", 
              financial_year AS "financialYear", company_code AS "companyCode", 
              mode, description, is_active AS "isActive" 
       FROM document_series WHERE id = $1 AND is_deleted = false`,
      [id]
    );
    if (checkRes.rows.length === 0) return res.status(404).json({ error: "Series not found" });
    const old = checkRes.rows[0];

    const fields: string[] = [];
    const values: any[] = [];
    let paramIdx = 1;

    const fieldsMapping: Record<string, string> = {
      name: "name",
      prefix: "prefix",
      suffix: "suffix",
      description: "description",
      mode: "mode"
    };

    for (const [key, col] of Object.entries(fieldsMapping)) {
      if (data[key] !== undefined) {
        fields.push(`${col} = $${paramIdx++}`);
        values.push(data[key]);
      }
    }

    if (data.currentNumber !== undefined) {
      fields.push(`current_number = $${paramIdx++}`);
      values.push(parseInt(data.currentNumber) || 0);
    }
    if (data.runningLength !== undefined) {
      fields.push(`running_length = $${paramIdx++}`);
      values.push(parseInt(data.runningLength) || 6);
    }
    if (data.resetRule !== undefined) {
      fields.push(`reset_rule = $${paramIdx++}`);
      values.push(data.resetRule);
    }
    if (data.financialYear !== undefined) {
      fields.push(`financial_year = $${paramIdx++}`);
      values.push(data.financialYear);
    }
    if (data.companyCode !== undefined) {
      fields.push(`company_code = $${paramIdx++}`);
      values.push(data.companyCode);
    }
    if (data.isActive !== undefined) {
      fields.push(`is_active = $${paramIdx++}`);
      values.push(data.isActive !== false);
    }

    let updated = { ...old, ...data };

    if (fields.length > 0) {
      values.push(id);
      await pool.query(
        `UPDATE document_series SET ${fields.join(", ")}, modified_at = CURRENT_TIMESTAMP WHERE id = $${paramIdx}`,
        values
      );
    }

    const logId = "NAL-" + Date.now().toString() + Math.floor(Math.random() * 1000);
    const operator = (req.headers["x-user-name"] as string) || "Admin";
    await pool.query(
      `INSERT INTO numbering_audit_logs (
        id, uuid, series_id, series_name, action, document_no, old_value, new_value, details, operator, is_active, is_deleted, created_at, modified_at
      ) VALUES ($1, uuid_generate_v4()::varchar, $2, $3, 'UPDATE', '-', $4, $5, $6, $7, true, false, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
      [
        logId,
        id,
        updated.name,
        JSON.stringify(old),
        JSON.stringify(updated),
        `Updated configuration parameters for series '${updated.name}'`,
        operator
      ]
    );

    res.json(updated);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

router.delete("/api/numbering/series/:id", async (req, res) => {
  const { id } = req.params;

  try {
    const checkRes = await pool.query(
      `SELECT name FROM document_series WHERE id = $1 AND is_deleted = false`,
      [id]
    );
    if (checkRes.rows.length === 0) return res.status(404).json({ error: "Series not found" });
    const name = checkRes.rows[0].name;

    await pool.query(
      `UPDATE document_series SET is_active = false, is_deleted = true, modified_at = CURRENT_TIMESTAMP WHERE id = $1`,
      [id]
    );

    const logId = "NAL-" + Date.now().toString() + Math.floor(Math.random() * 1000);
    const operator = (req.headers["x-user-name"] as string) || "Admin";
    await pool.query(
      `INSERT INTO numbering_audit_logs (
        id, uuid, series_id, series_name, action, document_no, old_value, new_value, details, operator, is_active, is_deleted, created_at, modified_at
      ) VALUES ($1, uuid_generate_v4()::varchar, $2, $3, 'RESET', '-', 'Active', 'Retired', $4, $5, true, false, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
      [
        logId,
        id,
        name,
        `Retired/deactivated series '${name}'`,
        operator
      ]
    );

    res.json({ success: true, message: `Series ${name} deactivated.` });
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

router.get("/api/numbering/logs", async (req, res) => {
  try {
    const dbRes = await pool.query(
      `SELECT id, created_at AS "timestamp", series_id AS "seriesId", 
              series_name AS "seriesName", action, operator AS "user", 
              document_no AS "documentNo", old_value AS "oldValue", 
              new_value AS "newValue", details 
       FROM numbering_audit_logs 
       ORDER BY created_at DESC`
    );
    res.json(dbRes.rows);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

router.post("/api/numbering/series/:id/allocate", async (req, res) => {
  const { id } = req.params;
  const pythonCoreHost = process.env.DATABASE_URL?.includes("@db:") ? "http://python-core:8000" : "http://localhost:8000";
  const authHeader = req.headers["authorization"];

  try {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-Internal-Service-Key": process.env.INTERNAL_SERVICE_KEY || "smriti_secret_fallback_key"
    };
    if (authHeader) {
      headers["Authorization"] = authHeader;
    }

    const payload = {
      branch: req.body.branch || "HQ",
      fy: req.body.fy || "26-27"
    };

    const upstreamRes = await fetch(`${pythonCoreHost}/api/v1/numbering/series/${id}/allocate`, {
      method: "POST",
      headers,
      body: JSON.stringify(payload)
    });

    if (!upstreamRes.ok) {
      const errText = await upstreamRes.text();
      try {
        const errJson = JSON.parse(errText);
        return res.status(upstreamRes.status).json({ error: errJson.detail || errJson.message || "Allocation failed" });
      } catch {
        return res.status(upstreamRes.status).json({ error: errText || "Allocation failed" });
      }
    }

    const data = await upstreamRes.json();
    return res.json(data);
  } catch (err: any) {
    return res.status(500).json({ error: err.message });
  }
});

export default router;

File: backend/app/api/v1/numbering.py
"""
Project      : SMRITI Retail OS
Author       : Jawahar Ramkripal Mallah
Designation  : Chief Systems Architect & Creator
Email        : support@smritibooks.com
Websites     : smritibooks.com | erpnbook.com | aitdl.com
Version      : 3.18.0
Created      : 2026-07-12
Modified     : 2026-07-12
Copyright    : © SMRITIBooks.com. All Rights Reserved.
License      : Proprietary Commercial Software
"""

from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from ...api.deps import get_db, get_current_user, require_role
from ...models.auth import User, UserRole
from ...schemas.numbering import (
    DocumentSeriesCreate, DocumentSeriesUpdate, DocumentSeriesResponse,
    NumberingAuditLogResponse, AllocationRequest,
    BillPrefixResolveRequest, BillPrefixResolveResponse,
    BillPrefixBatchSaveRequest, YearEndRolloverRequest, YearEndRolloverResponse
)
from ...services.numbering import NumberingService

router = APIRouter()


@router.get(
    "/series",
    response_model=List[DocumentSeriesResponse],
)
async def list_series(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    List all active document series configuration parameters.
    """
    service = NumberingService(db)
    return await service.list_series()


@router.post(
    "/series",
    response_model=DocumentSeriesResponse,
    status_code=201,
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def create_series(
    req: DocumentSeriesCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Create a new document series sequence scheme.
    """
    service = NumberingService(db)
    return await service.create_series(req, current_user.username)


@router.put(
    "/series/{id}",
    response_model=DocumentSeriesResponse,
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def update_series(
    id: str,
    req: DocumentSeriesUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Update document series sequence properties.
    """
    service = NumberingService(db)
    return await service.update_series(id, req, current_user.username)


@router.delete(
    "/series/{id}",
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def delete_series(
    id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Soft-retire/Deactivate document series sequence configuration.
    """
    service = NumberingService(db)
    await service.delete_series(id, current_user.username)
    return {"success": True, "message": f"Series retired/deactivated."}


@router.get(
    "/logs",
    response_model=List[NumberingAuditLogResponse],
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def list_logs(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Fetch number sequence allocation and adjustments history logs.
    """
    service = NumberingService(db)
    return await service.list_audit_logs()


from typing import Optional
from fastapi import Header

@router.post(
    "/series/{id}/allocate",
)
async def allocate_number(
    id: str,
    req: AllocationRequest,
    db: AsyncSession = Depends(get_db),
    x_internal_service_key: Optional[str] = Header(None, alias="X-Internal-Service-Key"),
    authorization: Optional[str] = Header(None),
):
    """
    Atomically allocate next serial sequence document number.
    """
    from ...core.config import settings
    username = "System"

    if x_internal_service_key and x_internal_service_key == settings.INTERNAL_SERVICE_KEY:
        # Internal service request authorized
        pass
    else:
        # Fall back to standard access token auth
        if not authorization or not authorization.startswith("Bearer "):
            raise HTTPException(
                status_code=401,
                detail="A valid access token or internal service key is required."
            )
        token = authorization.split(" ")[1]
        # Avoid circular imports
        from ...api.deps import get_current_user
        current_user = await get_current_user(token=token, db=db)
        username = current_user.username

    service = NumberingService(db)
    doc_no = await service.allocate_voucher_number(
        series_id=id,
        branch=req.branch or "HQ",
        fy=req.fy or "26-27",
        username=username
    )
    return {"success": True, "documentNo": doc_no}


# =========================================================================
# Shoper 9 Bill Prefix Endpoints
# =========================================================================

@router.get(
    "/bill-prefixes",
    response_model=List[DocumentSeriesResponse],
)
async def list_bill_prefixes(
    transaction_group: Optional[str] = None,
    terminal_id: Optional[str] = None,
    branch_id: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    List bill prefixes with optional group, terminal, and branch filters.
    """
    service = NumberingService(db)
    return await service.list_bill_prefixes(
        company_id=getattr(current_user, "company_id", None),
        branch_id=branch_id,
        transaction_group=transaction_group,
        terminal_id=terminal_id
    )


@router.post(
    "/bill-prefixes/resolve",
    response_model=BillPrefixResolveResponse,
)
async def resolve_bill_prefix(
    req: BillPrefixResolveRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Resolves the active Bill Prefix, sequence preview, and statutory GST Rule 46(b) validation
    for a POS terminal counter before transaction creation.
    """
    service = NumberingService(db)
    return await service.resolve_bill_prefix(
        company_id=getattr(current_user, "company_id", None),
        branch_id=req.branchId,
        terminal_id=req.terminalId or "COMMON",
        transaction_type=req.transactionType,
        bill_type=req.billType or "Product"
    )


@router.post(
    "/bill-prefixes/save-batch",
    response_model=List[DocumentSeriesResponse],
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def save_bill_prefixes_batch(
    req: BillPrefixBatchSaveRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Batch save or update document prefix schemes from the Prefix Management window.
    Enforces GST Rule 46(b) validation on each prefix scheme.
    """
    service = NumberingService(db)
    return await service.save_bill_prefixes_batch(
        company_id=getattr(current_user, "company_id", None),
        branch_id=req.branchId,
        req=req,
        operator=current_user.username
    )


@router.post(
    "/year-end-rollover",
    response_model=YearEndRolloverResponse,
    dependencies=[Depends(require_role(UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def execute_year_end_rollover(
    req: YearEndRolloverRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Supervisory Year End Process:
    Increments financial year suffix across all active document series,
    resets starting document numbers, and logs immutable audit trail.
    """
    service = NumberingService(db)
    return await service.execute_year_end_rollover(
        company_id=getattr(current_user, "company_id", None),
        req=req,
        operator=current_user.username
    )


@router.get(
    "/terminal-prefixes-report",
    dependencies=[Depends(require_role(UserRole.CASHIER, UserRole.MANAGER, UserRole.SYSADMIN))],
)
async def get_terminal_prefixes_report(
    terminal_id: Optional[str] = None,
    branch_id: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Terminal Prefix Listing report: detailed breakdown of bill prefixes defined per terminal node.
    """
    service = NumberingService(db)
    series_list = await service.list_bill_prefixes(
        company_id=getattr(current_user, "company_id", None),
        branch_id=branch_id,
        terminal_id=terminal_id
    )
    rows = []
    for s in series_list:
        pfx = s.prefix or ""
        sfx = s.suffix or ""
        next_n = (s.current_number or (s.start_number - 1)) + 1
        fmt = str(next_n).zfill(s.running_length or 4)
        preview = service._assemble_doc_no(
            pfx, fmt, sfx,
            s.financial_year,
            getattr(s, "number_format", None)
        )
        gst_eval = service.validate_gst_rule_46b(pfx, fmt, sfx)
        rows.append({
            "seriesId": s.id,
            "name": s.name,
            "terminalId": s.terminal_id or "COMMON",
            "isCommonAcrossTerminals": s.is_common_across_terminals if s.is_common_across_terminals is not None else True,
            "documentType": s.document_type,
            "transactionGroup": s.transaction_group or "SALES",
            "prefix": pfx,
            "suffix": sfx,
            "startNumber": s.start_number or 1,
            "currentNumber": s.current_number or 0,
            "nextDocumentNo": next_n,
            "preview": preview,
            "runningLength": s.running_length or 4,
            "financialYear": s.financial_year or "2026-2027",
            "isActive": s.is_active,
            "gstRule46bValid": gst_eval["isValid"],
            "gstRule46bLength": gst_eval["length"],
            "numberFormat": getattr(s, "number_format", None) or "PREFIX_NUM_SUFFIX",
        })
    return {"success": True, "count": len(rows), "items": rows}

